enhancements/ingress/set-delete-http-headers.md

```diff
@@ -0,0 +1,1031 @@
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=1024
+    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
+    // +kubebuilder:validation:XValidation:rule="self.lowerAscii() != 'strict-transport-security'",message="strict-transport-security may not be set/delete via header actions"
+    // +kubebuilder:validation:XValidation:rule="self.lowerAscii() != 'proxy'",message="proxy may not be set/delete via header actions"
+    // +kubebuilder:validation:XValidation:rule="self.lowerAscii() != 'host'",message="host may not be set/delete via header actions"
+    Name string `json:"name"`
+    // action specifies actions to perform on headers, such as setting or deleting headers.
+    // +kubebuilder:validation:Required
+    Action IngressControllerHTTPHeaderActionUnion `json:"action"`
+}
+```
+
+```go
+// IngressControllerHTTPHeaderActionUnion specifies an action to take on an HTTP header.
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Set' ?  has(self.set) : !has(self.set)",message="set is required when type is Set, and forbidden otherwise"
+// +union
+type IngressControllerHTTPHeaderActionUnion struct {
+    // type defines the type of the action to be applied on the header.
+    // Possible values are Set or Delete.
+    // Set allows you to set http request and response headers.
+    // Delete allows you to delete http request and response headers.
+    // +unionDiscriminator
+    // +kubebuilder:validation:Enum:=Set;Delete
+    // +kubebuilder:validation:Required
+    Type IngressControllerHTTPHeaderActionType `json:"type"`
+
+    // set specifies how the HTTP header should be set.
+    // If this field is empty, no action is taken.
+    // +optional
+    // +unionMember
+    Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`
+}
+```
+
+```go
+// IngressControllerHTTPHeaderActionType defines actions that can be performed on HTTP headers.
+type IngressControllerHTTPHeaderActionType string
+
+const (
+	// Set specifies that an HTTP header should be set.
+	Set IngressControllerHTTPHeaderActionType = "Set"
+	// Delete specifies that an HTTP header should be deleted.
+	Delete IngressControllerHTTPHeaderActionType = "Delete"
+)
+```
+
+The `IngressControllerSetHTTPHeader` type has a field for specifying a header value.
+```go
+// IngressControllerSetHTTPHeader defines the value which needs to be set on an HTTP header.
+type IngressControllerSetHTTPHeader struct {
+	// value specifies a header value.
+	// Dynamic values can be added. The value will be interpreted as an HAProxy format string as defined in
+	// https://cbonte.github.io/haproxy-dconv/2.4/configuration.html#8.2.4 and may use HAProxy's %[] syntax and
+	// otherwise must be a valid HTTP header value as defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.
+	// It must be no more than 32768 characters in length.
+	// This value was selected as most of the servers accept this value.
```

Comment:
Please add a note that the total size of all added headers _after_ interpolating dynamic values must not exceed the rewrite buffer size.  Please add the note in these four places for visibility:
* The `value` field of `IngressControllerSetHTTPHeader` (i.e. here).
* The `value` field of `RouteSetHTTPHeader`.
* The `actions` field of `IngressControllerHTTPHeaders`.
* The `actions` field of `RouteHTTPHeaders`.
```suggestion
	// The value of this field must be no more than 32768 characters in length.
	// This limit was selected as most common web servers have a limit of 16384 characters
	// or some lower limit.
	//
	// Note that the total size of all added headers *after* interpolating dynamic values
	// must not exceed the value of spec.tuningOptions.headerBufferMaxRewriteBytes on the
	// IngressController.
```